api/v1alpha1/provisioning_types.go

```diff
@@ -99,6 +99,10 @@ type ProvisioningSpec struct {
 	// accessible from the machine networks. User should provide two IPs on
 	// the external network that would be used for provisioning services.
 	ProvisioningNetwork ProvisioningNetwork `json:"provisioningNetwork,omitempty"`
+
+	// EnableMultiNamespaces provides a way to explicitly allow use of this
+	// Provisioning configuration across multiple namespaces.
+	EnableMultiNamespaces bool `json:"enableMultiNamespaces,omitempty"`
```

Comment:
I wonder if we can use some existing API for this, e.g for BMH resources we already set metadata/namespace:

https://github.com/openshift/installer/blob/master/pkg/asset/machines/baremetal/hosts.go#L75

However, we unfortunately didn't add that to the provisioning CR here:

https://github.com/openshift/installer/blob/master/data/data/manifests/openshift/baremetal-provisioning-config.yaml.template#L4

But I think such an interface might fit more cleanly, e.g similar to the OLM operator namespace behavior described here:

https://docs.openshift.com/container-platform/4.6/operators/understanding/olm/olm-understanding-operatorgroups.html#olm-operatorgroups-target-namespace_olm-understanding-operatorgroups

Is there any way we could add that missing metadata on upgrade, e.g by looking at the status/generations which does show the openshift-machine-api namespace?

Alternatively if we think modifying that existing behavior will be too difficult or ambiguous, I'd suggest renaming this interface to something terser e.g:

```
   spec:
        allNamespaces: true
```

That would align pretty well with the OLM docs e.g `targetNamespaces` ?

Reply:
Currently, Provisioning CR is cluster scoped. If we change it to Namespace scope, we would get this metadata change. This would be a breaking API change during an upgrade. The solution that is proposed here only tells us if multi-namespace mode is enabled. It will not allow up to specify an actual namespace. When we move to model where there would be a Provisioning CR per namespace, we could make the change to make Provisioning CRD with its scope set to Namespace.

Reply:
After going through the links mentioned above, I see that the semantics of the OLM operator namespace behavior different from what we are looking for here. If we re-use the name, would it cause some confusion?

Reply:
@sadasu Ok sounds like we should go with the existing boolean, but perhaps rename the flag to `allNamespaces` (not blocking on that, but it seems to communicate the same as the proposed flag in a slightly terser and more accurate way, since "multi" could mean all, or some subset of all?)

Reply:
- Would we never want to watch some but not all namespaces?

- I'm also not a fan of more knobs.  If it's always going to be turned on, why do we need it? And if it can be conditionally on/off how does one do that from the installer? Or do you change it on day 2?
-

Reply:
When in this mode we ask BMO to watch all Namespaces (by not setting the WATCH_NAMESPACE env var), although in practice BMH objects would likely be created in some Namespaces.

Point taken about additional knob. Best solution would be to make Provisioning CR Namespaced but that is coming later. This config would be false by default and also from the Installer and can be set to true on Day2.

Also, @kirankt 's suggestion to call the knob, "WatchAllNamespaces" is a better choice that what I have now, imo. I am not sure if it is terse enough but it conveys its purpose clearly.

Reply:
>   Would we never want to watch some but not all namespaces?

Possibly, but I don't think there's any requirement for that right now.

>  I'm also not a fan of more knobs. If it's always going to be turned on, why do we need it? And if it can be conditionally on/off how does one do that from the installer? Or do you change it on day 2?

IMO this shouldn't always be turned on, particularly given that we want to align with the machine-api e.g "In a normal OpenShift cluster we restrict the controllers to only watch a single namespace (openshift-machine-api)" comment from @JoelSpeed 

The future direction for multi-namespace at the machine-api layer seems unclear at this point, so it seems safest to leave the existing single-namespace behavior alone, and add a knob to enable the specific "hub cluster" use-case which isn't a normal OpenShift cluster.

In terms of how to configure it, I was thinking either modify the installer manifest or ensure it can be turned on after initial deployment (at least until the story around multi-namespace machine management is worked out for other controllers, when perhaps there will be some kind of installer flag).